File: engine/runtime/src/timer.cpp
#include "timer.h"
#include <SDL3/SDL_timer.h>

namespace runa::runtime
{
    time_c::time_c()
    {
    }

    time_c::~time_c()
    {
    }

    void time_c::update_current_time()
    {
        current_time_ns = SDL_GetTicksNS();
    }

    void time_c::update_end_time()
    {
        delta_time_ns = SDL_GetTicksNS() - current_time_ns;
    }

    uint64_t time_c::elapsed_ns()
    {
        return SDL_GetTicksNS() - current_time_ns;
    }

    uint64_t time_c::delta_ns()
    {
        return delta_time_ns;
    }

    double time_c::delta()
    {
        return double(delta_time_ns) / 1000000000.0;
    }

    time_c Time = time_c();
}

File: engine/runtime/timer.zig
const runtime = @import("runtime");
const sdl = @import("sdl3");

pub const Time = struct {
    currentTimeNS: u64,
    deltaTimeNS: u64,

    pub fn init() Time {
        var self: Time = undefined;
        self.currentTimeNS = 0;
        self.deltaTimeNS = 0;

        return self;
    }
    pub fn updateCurrentTime(self: *Time) void {
        self.currentTimeNS = sdl.timer.getNanosecondsSinceInit();
    }
    pub fn updateEndTime(self: *Time) void {
        self.deltaTimeNS = sdl.timer.getNanosecondsSinceInit() - self.currentTimeNS;
    }
    pub fn elapsedNS(self: *Time) u64 {
        return sdl.timer.getNanosecondsSinceInit() - self.currentTimeNS;
    }
    pub fn deltaNS(self: *Time) u64 {
        return self.deltaTimeNS;
    }
    pub fn delta(self: *Time) f64 {
        return @as(f64, @floatFromInt(self.deltaTimeNS)) / 1_000_000_000.0;
    }
};
